client/src/auth/PrivateRoute.jsx

```diff
@@ -0,0 +1,71 @@
+import React, { Component } from "react";
+import { Route, withRouter } from "react-router-dom";
+import axios from 'axios';
+
+class PrivateRoute extends Component {
+  initialMount = false;
+
+  // axios configuration
+  axiosSource = axios.CancelToken.source();
+  axiosInstance = axios.create({
+    baseURL: '/api/v1/',
+    cancelToken: this.axiosSource.token
+  });
+
+  state = {
+    isAuth: false,
+  }
+
+  componentWillUnmount() {
+    this.axiosSource.cancel('cancel isAuthenticated request');
+  }
+
+  componentDidMount() {
+    this.axiosInstance
+    .get('/isAuthenticated')
+      .then(() => {
+        this.initialMount = true;
+        this.setState({ isAuth: true });
+      })
+      .catch(() => {
+        this.isUnAuthorized();
+      });
+  }
+
+  componentDidUpdate() {
+    // this will stop when updating after the initial mount
```

Comment:
@ashatat I think the componentDidMount() logic will handle setting state.isAuth when the component initially mounts (and so you don't need this componentDidUpdate()).  Was there a specific edge case that you needed this componentDidUpdate() logic?

Reply:
We want to make sure the user is still authenticated, meaning he does not play with the cookie each time he navigates the web. This is why we use componentDidUpdate() in the PrivateRoute Component. Does this make any sense, @tomduggan85?

Reply:
:grimacing:  it was just kind of challenge to Authenticate (from the server side) before the user does any changes Also when he change the routes, so PrivaterRoute didn't get mounted when you change the routes.

To be honest, I am not sure if it makes sense to Authenticate each time, I know that Authentication and security should be on the server side, (the client side code is exposed to everyone)

I am not sure if that necessary or not, I would like to know what do you think about it

Also, there is the `reactrouter` example of authentication where they don't rely on the state for Authentication, they change a property inside Object.
[here is the link](https://codesandbox.io/s/1q9z6yz87)

Reply:
Oh I see, so if you view a private route once (and componentDidMount() decides that you are authenticated), then you log out of the app, then visit that same route again, it does not re-mount so it still considers you authenticated?
I hadn't thought of that - In that case maybe you do need something like this.  I wonder if you could check this.props.location.pathname !== prevProps.location.pathname in componentDidUpdate, to re-run authentication only when the user changes pages.
But then if you have (for example) 5 routes, all 5 routes are making a network request to /isAuthenticated when the app starts, and all 5 routes would again fetch /isAuthenticated whenever the page changes, which is a lot of network requests.

So given all that, maybe it does make more sense to (like I had suggested as a next improvement) move all of the responsibility for this - fetching /isAuthenticated and storing a `state.isLoggedIn` - into App.js, and then pass isLoggedIn into each LoggedOutRoute / PrivateRoute as a prop.  Then maybe you would avoid issues like these entirely.  PrivateRoute would simply render something like this:
`this.props.isLoggedIn ? <Component {...rest} /> : <Redirect to='/login' />`
.. and wouldn't need any logic in its componentDidMount / componentDidUpdate.
It might be too late into the project life to make another architectural change like that, so it's totally up to you - whatever works best for you.

Regarding authentication on the server side vs. client side and what is necessary - I think what you are doing is good, and is a pattern I see a lot.  The server will ultimately be in charge of authentication - it won't return data if you don't make an authenticated request - but having a little extra logic in the client, to redirect the client back to the login page if they haven't been on your website in 2 weeks instead of letting them go to /projects and just have broken network requests, is a nicer user experience!

Reply:
@IsraaSulaiman Ah yes that makes sense.

Reply:
Thanks, @tomduggan85 Your explanation is so useful.

Reply:
I am still thinking about what you said. Why a user would have a broken network request if they go to /projects after two weeks? The cookie will be cleared, so when a user requests projects route, the private route will send a request to check if the user is authenticated or not. In this case, the user will not be authenticated, so he will be redirected to login route. (bah...I have authentication)

Reply:
@IsraaSulaiman Authentication is lovely :stuck_out_tongue: 
But I think we make it complicated here :man_facepalming:
maybe we should stop the didUpdate check, and move the authentication check in the app as @tomduggan85  said, then we rely on the error handling for each request we sent to fetch data from the API, if it returns 401 then we should send him a warning then return him back to the lovely login page :stuck_out_tongue_closed_eyes:

it's just a thought :grimacing: 

@tomduggan85   **Really Thank you for your explanation, I forgot that today is your weekend, SORRY**   :grimacing:

Reply:
@IsraaSulaiman you’re welcome! When I said the 2 week (or 1 week rather) thing, I meant that would be the user experience if you didn’t have your client-side isAuthenticated fetch, and only relied on the server authentication that occurs on all other fetches (for example fetching /api/projects). In other words, I think your client-side isAuthenticated logic is a better user experience :)
Authentication is hard but at least you only have to write it / worry about it once for each app you make :)
@ashatat no problem! I’m happy to help even on weekends. I think that plan might help make the auth less complicated - App gets isAuthenticated once on mount, puts it into state, updates the state whenever the user does a login or logout, and then passes that login state down as a prop to each PrivateRoute or LoggedOutRoute. Then it will only be fetching isAuthenticated once for every App session. And yeah you can redirect to login upon getting a 401 from any network request (for this reason, some projects create their own api function, an axios/fetch wrapper, so they can have a place to put that 401 handling)